Add `Node.get_sibling_lists` method

jwallwork23 opened the pull request:
Closes #2370.

Given some node (e.g., the schedule for an invoke), the proposed method returns lists of nodes of a specified type (or tuple of types) such that:
* All nodes within a list have the same `depth`.
* The nodes within the list follow one another.

As mentioned in the issue, the motivation here is OpenACC GPU porting, where it's beneficial to put adjacent constructs into the same `kernels` region.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
             local_list += child.walk(my_type, stop_type)
         return local_list
 
+    def split_consecutive(self, my_type, stop_type=None):
```
I'm not sure whether the name `split_consecutive` is the best choice. The operation does feel like Python string splitting but the name doesn't indicate that there will be any `walk`ing involved. I also considered a name including "twins" (immediate siblings). Open to suggestions!

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
 
+    def split_consecutive(self, my_type, stop_type=None):
+        '''
+        Recurse through the PSyIR tree and return lists of objects that are
+        instances of 'my_type' and immediately follow one another. Here
+        'my_type' is either a single class or a tuple of classes. In the latter
+        case all nodes are returned that are instances of any classes in the
+        tuple. The recursion into the tree is stopped if an instance of
+        'stop_type' (which is either a single class or a tuple of classes) is
+        found.
```
Docstring largely copied from that of `walk` - hopefully that's okay.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1284,6 +1327,30 @@ def preceding(self, reverse=False, routine=True):
             nodes.reverse()
         return nodes
 
+    def immediatelyPrecedes(self, node_2):
```
Happy to change name to `immediately_precedes` if preferred (I was following `sameParent`).

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1284,6 +1327,30 @@ def preceding(self, reverse=False, routine=True):
+            and self.position + 1 == node_2.position
+        )
+
+    def immediatelyFollows(self, node_1):
```
Similarly, happy to change to `immediately_follows`.

jwallwork23 replied:
Done, see e8bf615efb4f3979e8ee830bc700afd6a5b25d8d.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2226,3 +2226,41 @@ def test_walk():
+        assert len(block) == (4 if i == 3 else 1)
+
+
+def test_split_consecutive2():
```
I prefer naming rather than numbering as it makes life easier in the long run - please could you rename this test to make it a bit more meaningful.

jwallwork23 replied:
Addressed in f65ee7f25436497e6a991a0c79f3047f154d93a1.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2226,3 +2226,41 @@ def test_walk():
+    '''Tests the split_consecutive functionality for
+    explicit_do_two_loops.f90.'''
+
+    _, invoke = get_invoke("explicit_do_two_loops.f90", "nemo", 0)
```
Again, when you move this test, please could you re-write to use `fortran_reader`.

jwallwork23 replied:
Done. See f65ee7f25436497e6a991a0c79f3047f154d93a1.

arporter commented on src/psyclone/tests/psyir/nodes/node_test.py:
```diff
@@ -1481,6 +1482,16 @@ def test_following_preceding():
     assert multiply1.following() == [c_ref, d_ref]
     assert multiply1.preceding() == [assign1, a_ref, multiply2, b_ref]
 
+    # 1d: Immediately following
```
Please could you extend slightly just to cover the case where `immediately_follows` and `immediately_precedes` return False.

jwallwork23 replied:
Ah yes good point. See ec05c0403c40616d7d0d4390933a53a26bad7a15.

arporter reviewed (changes requested):
Nice functionality, thanks for this. Just a bit of renaming and tidying to do.

We should also be documenting these new methods.  Would you mind adding a section similar to 'Tree Navigation' (https://psyclone.readthedocs.io/en/stable/psyir.html#tree-navigation) named 'Tree Interrogation' to the User Guide. It will just need a similar bit of preamble and then you can use the Sphinx autodoc feature to pull in the documentation of these new methods from their docstrings c.f. what is done in 'Tree Navigation'.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
+                else:
+                    global_list.append(block)
+                    block = [node]
+            global_list.append(block)
```
Good point. See 1ba9be0b26ee87d76085ec39c7e3e6cde99b7d63.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
+        for depth, local_list in sorted(by_depth.items()):
+            block = []
+            for node in local_list:
+                if len(block) == 0 or node.immediatelyFollows(block[-1]):
```
Fixed in 125e72d432a9b03e43928b9954383f8f52659c10.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
+                  starting at and including this node.
+        :rtype: List[List[:py:class:`psyclone.psyir.nodes.Node`]]
+        '''
+
```
Done. See 515be1a6b62539a8612197c0a81d62b340ccf90a.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
 
+    def split_consecutive(self, my_type, stop_type=None):
+        '''
+        Recurse through the PSyIR tree and return lists of objects that are
```
Done. See cb01bfc16b9f7a9caf07494a4b244381f34b098a.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
+    def split_consecutive(self, my_type, stop_type=None):
+        '''
+        Recurse through the PSyIR tree and return lists of objects that are
+        instances of 'my_type' and immediately follow one another. Here
```
Done. See 42d311fcb262584afe1378edb9e3bb3df4352373.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
+                if len(block) == 0 or node.immediatelyFollows(block[-1]):
+                    block.append(node)
+                else:
+                    global_list.append(block)
```
Suggested comment added in b1429d6e6052fda550f5a8f62baf9a995ba06a31.

jwallwork23 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1094,6 +1095,48 @@ class or a tuple of classes) is found. This can be used to avoid
+        :param stop_type: class(es) at which recursion is halted (optional).
+        :type stop_type: Optional[type | Tuple[type, ...]]
+
+        :returns: list of lists, each of which containing nodes that are \
```
Done. See 4cbf2b3d4506b127a1bd804476e8728469ac6ba2.

jwallwork23 commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2226,3 +2226,41 @@ def test_walk():
     assert not binary_op_list
+
+
+def test_split_consecutive1():
```
Addressed this in f65ee7f25436497e6a991a0c79f3047f154d93a1. I completely reworked the tests, with similarly careful checks as in #2368.

jwallwork23 commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2226,3 +2226,41 @@ def test_walk():
+def test_split_consecutive1():
+    '''Tests the split_consecutive functionality for imperfect_nest.f90.'''
+
+    _, invoke = get_invoke("imperfect_nest.f90", "nemo", 0)
```
Done. See f65ee7f25436497e6a991a0c79f3047f154d93a1.

jwallwork23 commented:
> We should also be documenting these new methods. Would you mind adding a section similar to 'Tree Navigation' (https://psyclone.readthedocs.io/en/stable/psyir.html#tree-navigation) named 'Tree Interrogation' to the User Guide. It will just need a similar bit of preamble and then you can use the Sphinx autodoc feature to pull in the documentation of these new methods from their docstrings c.f. what is done in 'Tree Navigation'.

Hopefully what I added in 5fc06aaf793ec512ddab0e89d19a444e583eb440 is along the lines of what you're looking for?

jwallwork23 commented:
I think I addressed all the comments. I also added another test to cover the way we have actually been using `get_sibling_lists` in practice, i.e. to identify blocks of code to put OpenACC `kernels` directives around. (In that case, we just want the outer-most blocks of the specified types, rather than all such blocks.)

arporter reviewed (changes requested):
All requested changes have been made (thanks) and new documentation is good :-)
However, now I look at this a second time I'm wondering whether it would be more natural to have `get_sibling_lists` return lists of Nodes rather than indices (because I found myself wondering what the indices were with respect to). Presumably that would then make it easier to apply subsequent transformations? Am happy to discuss of course.

As this is a small PR it doesn't matter here but for future reference if you could avoid force pushing that would really help the review process (as we tend to rely on being able to see the changes made since the last review). Thanks :-)

arporter commented on doc/user_guide/psyir.rst:
```diff
@@ -219,6 +220,28 @@ tree from an ancestor node to the node:
+Each PSyIR node provides several ways to interrogate the AST:
+
+Following the `parent` and `children` terminology, we define a node's `siblings`
+as the children of its parents. Note that this definition implies that all nodes
```
As I've returned this to you, would you also mind replacing 'parents' with 'parent' here (as a node only has a single parent) and you get the glory (?)  of being able to add your name to the author list (conf.py).

jwallwork23 replied:
Done, see 2f1afecb0.

jwallwork23 replied:
Sorry @arporter, which part of the `conf.py` file should I add to? There seem to be three different author lists.

arporter replied:
Oh yes, all three of them please!

jwallwork23 replied:
Oh actually there were another two. I think this were the right places? 15c0b1a13

jwallwork23 commented:
> As this is a small PR it doesn't matter here but for future reference if you could avoid force pushing that would really help the review process (as we tend to rely on being able to see the changes made since the last review). Thanks :-)

Yes, I can do that going forward. (I've worked on codes in the past with a preference for rebasing.)

jwallwork23 commented:
> I'm wondering whether it would be more natural to have `get_sibling_lists` return lists of Nodes rather than indices (because I found myself wondering what the indices were with respect to). Presumably that would then make it easier to apply subsequent transformations? Am happy to discuss of course.

Perhaps I've misunderstood your question, but the current version of `get_sibling_lists` does return a list of Nodes, doesn't it? Once it's established a block of Nodes as `local_list`, it adds it to `global_list`, which is later returned.

arporter commented:
> > As this is a small PR it doesn't matter here but for future reference if you could avoid force pushing that would really help the review process (as we tend to rely on being able to see the changes made since the last review). Thanks :-)
> 
> Yes, I can do that going forward. (I've worked on codes in the past with a preference for rebasing.)

Thank you :-) You generally don't have to rebase - a merge is often fine (and seems to be simpler).

arporter reviewed (approved):
All requested changes have been made (well, those that were sensible).
CI is green.
Integration tests are green.
Will proceed to merge.
